### ISSUE TITLE
Upload processes all files and does not delete remot files

### ISSUE THREAD

**Issue by xannasavin** [SamKirkland/FTP-Deploy-Action#113]
**Bug Description**
Hey,
I'm just getting started with those github actions and found your action the best fitting after trying a few.
However, I'm experiencing some problems, that could also come from inexperience, sorry if that's the case.

I'm uploading my local Wordpress themes and plugins folders through github to my prod system, the scope is the whole wp-contents folder, ignoring everything besides themes and plugins.

Every run takes over an hour, since it looks like all files are checked and maybe even re-uploaded - not sure about the later.
Also files that I deleted on local are deleted on github but remain on the prod system.

If any more information needed, please let me know.

Thanks a lot in advance!

**My Action Config**
```yaml
# This is a basic workflow to help you get started with Actions

name: CI for deployment to PROD

# Controls when the action will run. Triggers the workflow on push or pull request
# events but only for the main branch
on:
  push:
    branches: [ main ]
  # pull_request:
    # branches: [ main ]

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "build"
  build:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
      - uses: actions/checkout@v2

      # FTP Deployment
      - name: FTP Deploy
        # You may pin to the exact commit or the version.
        # uses: SamKirkland/FTP-Deploy-Action@da0d77ff390a13097c60f9b9eb954e503a480d97
        # https://github.com/marketplace/actions/ftp-deploy
        uses: SamKirkland/FTP-Deploy-Action@3.1.1
        with:
          # Deployment destination server & path. Formatted as protocol://domain.com:port/full/destination/path/
          ftp-server: ${{ secrets.FTP_SERVER }}
          # FTP account username
          ftp-username: ${{ secrets.FTP_USER }}
          # FTP account password
          ftp-password: ${{ secrets.FTP_PASSWORD }}
          # The local folder to copy, defaults to root project folder
          #local-dir: # optional, default is ./
          # Passes through options into git-ftp
          git-ftp-args: --lock # optional
```

**My Action Log**
```
   Log would be too long - in general it runs well, but through all files without deleting. I can attach if necessary
```

.gitignore in root folder:
```
.idea/*
.idea/
.idea/modules.xml
.idea/vcs.xml
.idea/workspace.xml
/upgrade/
/uploads/
!themes/*
!plugins/*
/languages/
.ablage
/updraft/
/cache/
/plugins/duplicator/
.github/workflows/deployment_old.yaml
```

.git-ftp-ignore in root folder
```
# Git Files
.github/*
.gitignore
*/.gitignore
*/.gitkeep
.git-ftp-ignore
.git-ftp-include
.gitlab-ci.yml
readme.md
*.md
# WP Files
/logs/*
/upgrade/*
/uploads/*
/languages/*
/themes/groupfinder_s/env.php
/updraft/*
/cache/*
/themes/groupfinder_s/dev_src/
```

**Comment by xannasavin** [comment]
I found this issue here https://github.com/SamKirkland/FTP-Deploy-Action/issues/6 which says the script should only upload changed files - so it does it somehow check GIT what has changed or does it scan all files on the target and check if the files in GIT are newer? that would explain why it takes over an hour for the script to run..

still desperate for answers please :)

**Comment by SamKirkland** [comment]
Hey!

Let me try and clear some stuff up. This library (version 3 at least) uses your git history to figure out what has changed between this publish and the last. Then it only uploads the newer items.

In your config you are only getting the **most recent** version of the code, so it appears to the script as if this is your first commit.

**Fix #1:** Update your checkout step action to match the [usage example](https://github.com/SamKirkland/FTP-Deploy-Action#usage-example)** this number needs to be at least 2. If you don't plan on deploying every checkin you can increase the number without much of a performance problem.
```yml
    - uses: actions/checkout@v2.1.0
      with:
        fetch-depth: 100
        # get the last XXX commits from history
        # Make sure this number is AT LEAST 2, that's because we need to see what has changed.
```

**Fix #2:** I'm working on changing how this diffing logic works to no longer have this limitation. You can use the [v4 beta](https://github.com/SamKirkland/FTP-Deploy-Action/tree/beta-v4) and after the first deployment you should no longer run into this problem.

--------

Let me know if that fixes it!

**Comment by xannasavin** [comment]
That did the job - thanks a lot! <3 
Last run was 7min, though a bunch of stuff changed. 
Looking forward to v4 :) 
Have a great day
